src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlSettingsAction.java

```diff
@@ -0,0 +1,122 @@
+import org.elasticsearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.elasticsearch.client.Requests;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.elasticsearch.rest.RestStatus.INTERNAL_SERVER_ERROR;
+
+/**
+ * Interface to manage opendistro.sql.* cluster settings
+ * All non-sql settings are ignored.
+ * Any non-transient and non-persistent settings are ignored.
+ */
+public class RestSqlSettingsAction extends BaseRestHandler {
+    private static final Logger LOG = LogManager.getLogger(RestSqlSettingsAction.class);
+
+    private static final String PERSISTENT = "persistent";
+    private static final String TRANSIENT = "transient";
+    private static final String SQL_SETTINGS_PREFIX = "opendistro.sql.";
+
+    /**
+     * API endpoint path
+     */
+    public static final String SETTINGS_API_ENDPOINT = "/_opendistro/_sql/settings";
+
+    public RestSqlSettingsAction(Settings settings, RestController restController) {
+        super();
+        restController.registerHandler(RestRequest.Method.POST, SETTINGS_API_ENDPOINT, this);
+    }
+
+    @Override
+    public String getName() {
+        return "sql_settings_action";
+    }
+
+    /**
+     * @see org.elasticsearch.rest.action.admin.cluster.RestClusterUpdateSettingsAction
+     */
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        LogUtils.addRequestId();
+        final ClusterUpdateSettingsRequest clusterUpdateSettingsRequest = Requests.clusterUpdateSettingsRequest();
+        clusterUpdateSettingsRequest.timeout(request.paramAsTime("timeout", clusterUpdateSettingsRequest.timeout()));
+        clusterUpdateSettingsRequest.masterNodeTimeout(
+                request.paramAsTime("master_timeout", clusterUpdateSettingsRequest.masterNodeTimeout()));
+        Map<String, Object> source;
+        try (XContentParser parser = request.contentParser()) {
+            source = parser.map();
+        }
+
+        try {
+            if (source.containsKey(TRANSIENT)) {
+                clusterUpdateSettingsRequest.transientSettings((Map) source.get(TRANSIENT));
+            }
+            if (source.containsKey(PERSISTENT)) {
+                clusterUpdateSettingsRequest.persistentSettings((Map) source.get(PERSISTENT));
+            }
+
+            // filter out all non-sql settings
+            clusterUpdateSettingsRequest.transientSettings(
```

Comment:
why call transientSettings two times? ref: line 86.

Reply:
This is to take advantage of [`ClusterUpdateSettingsRequest.transientSettings(Map<String, ?> source)`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/admin/cluster/settings/ClusterUpdateSettingsRequest.java#L117) method, instead of redoing again. The second `transientSettings()` call just assigns new Settings after filtering non `opendistro.sql` settings.


We can bring the same logic from [`ClusterUpdateSettingsRequest.transientSettings(Map<String, ?> source)`](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/admin/cluster/settings/ClusterUpdateSettingsRequest.java#L117), if there is a performance impact.

Reply:
No performance impact, I have a little bit concern that the current implementation coupled with ClusterUpdateSettingRequest internal implementation. It could be pain for long term maintenance.
As you said, maybe we can use copy the logic and only use the public interface if possible.

Reply:
I'm thinking should we add an abstraction for this rather than exposing our internal implementation to customer directly. For example,

```
POST _opendistro/_sql/settings
{
    "defaultReponseFormat": "JSON"
}
```

Reply:
That works fine as well. Will change as suggested.

Reply:
Just my thought and open for discussion. Not sure if any potential problem and security concern with the current implementation. Just feel it would be safer and better to hide our implementation.

Reply:
My thoughts:

1. Yes we can abstract, but I don't see any major advantage of doing this. In terms of safety, we are already filtering based on `opendistro.sql.` prefix. Since the plugin is  open source, hiding implementation would be just security by obscurity.

2. On top of that, there would be major work in changing the response coming from backend, to make request/response consistent. These most likely will affect existing clients which are parsing the old response from `_cluster/settings` API.

Reply:
> No performance impact, I have a little bit concern that the current implementation coupled with ClusterUpdateSettingRequest internal implementation. It could be pain for long term maintenance.
> As you said, maybe we can use copy the logic and only use the public interface if possible.

Done.

Reply:
@abbashus Got it. Do you know if this would work well with security plugin? Understood that the impact would be limited to SQL plugin only, but wonder if a customer without enough permission, let's say disable SQL plugin, can do it and impact others.

Reply:
> @abbashus Got it. Do you know if this would work well with security plugin? Understood that the impact would be limited to SQL plugin only, but wonder if a customer without enough permission, let's say disable SQL plugin, can do it and impact others.

Let me deep dive on this and confirm how this will work in tandem with security plugin, and if there are any security implications or not.

Reply:
@dai-chen I verified the behavior with security plugin. A user which does not have `cluster:admin/settings/update` permission cannot modify SQL settings by using new endpoint as underneath both requires same permissions.

Reply:
Thanks a lot for verifying!

Reply:
Adding here for record , how the user will be denied changing SQL settings in absence of `cluster:admin/settings/update` permissions.


```
PUT /_opendistro/_sql/settings?pretty 
{
  "transient": {
    "opendistro.sql.query.slowlog": "20"
  }
}

{
  "error" : {
    "root_cause" : [
      {
        "type" : "security_exception",
        "reason" : "no permissions for [cluster:admin/settings/update] and User [name=abbas, backend_roles=[], requestedTenant=__user__]"
      }
    ],
    "type" : "security_exception",
    "reason" : "no permissions for [cluster:admin/settings/update] and User [name=abbas, backend_roles=[], requestedTenant=__user__]"
  },
  "status" : 403
}

```